joinmarket/peertopeer.py

```diff
@@ -0,0 +1,307 @@
+        return base64.b32encode(ip_hex[6:]).lower() + '.onion'
+    else:
+        return socket.inet_ntop(socket.AF_INET6, ip_hex)
+
+class P2PMessageHandler(object):
+    def __init__(self):
+        pass
+
+    def handle_message(self, p2p, command, length, payload):
+        ptr = [0]
+        if command == 'version':
+            version = read_int(ptr, payload, 4)
+            services = read_int(ptr, payload, 8)
+            timestamp = read_int(ptr, payload, 8)
+            addr_recv_services = read_int(ptr, payload, 8)
+            addr_recv_ip = payload[ptr[0] : ptr[0]+16]
+            ptr[0] += 16
+            addr_recv_port = read_int(ptr, payload, 2, False)
+            addr_trans_services = read_int(ptr, payload, 8)
+            addr_trans_ip = payload[ptr[0] : ptr[0]+16]
+            ptr[0] += 16
+            addr_trans_port = read_int(ptr, payload, 2, False)
+            ptr[0] += 8 #skip over nonce
+            user_agent = read_var_str(ptr, payload)
+            start_height = read_int(ptr, payload, 4)
+            if version > RELAY_TX_VERSION:
+                relay = read_int(ptr, payload, 1) != 0
+            else:
+                relay = True
+            log.debug(('peer version message: version=%d services=0x%x'
+                + ' timestamp=%s user_agent=%s start_height=%d') % (version,
+                services, str(datetime.datetime.fromtimestamp(timestamp)),
+                user_agent, start_height))
+            log.debug('their addr = ' + ip_hex_to_str(addr_trans_ip) + ':'
+                + str(addr_trans_port) + ' our address according to them = '
+                + ip_hex_to_str(addr_recv_ip) + ':' + str(addr_recv_port))
+            p2p.sock.sendall(p2p.create_message('verack', ''))
+        elif command == 'verack':
+            log.debug('connected to peer')
+            self.on_connected(p2p)
+        elif command == 'ping':
+            p2p.sock.sendall(p2p.create_message('pong', payload))
+
+    def on_connected(self, p2p):
+        pass
+
+class P2PProtocol(object):
+    def __init__(self, p2p_message_handler, remote_hostport=None,
+            testnet=False, user_agent=DEFAULT_USER_AGENT, relay_txes=False,
+            socks5_hostport=None):
+        '''
+        if remote_hostport = None, use dns_seeds for auto finding peers
+        if socks5_hostport != None, use that proxy 
+        relax_txes controls whether the peer will send you unconfirmed txes
+        '''
+        self.p2p_message_handler = p2p_message_handler
+        self.testnet = testnet
+        self.user_agent = user_agent
+        self.relay_txes = relay_txes
+        self.socks5_hostport = socks5_hostport
+        if not self.testnet:
+            self.magic = 0xd9b4bef9 #mainnet
+        else:
+            if testnet == True:
+                self.magic = 0x0709110b #testnet
+            else:
+                self.magic = 0xdab5bffa #regtest
+        self.closed = False
+        self.connection_attempts = 4
```

Comment:
Increase this number to about 20 or 30 since 4 is far too low, also make it a defined constant MAX_CONNECTION_ATTEMPTS